### PR TITLE
hypre: Add ROCm variant

### PR DESCRIPTION
**Pull request by osborn9**
Use ROCmPackage to add ROCm variant to hypre package

### CODE REVIEW

File: var/spack/repos/builtin/packages/hypre/package.py [hunk 1]
```diff
@@ -103,6 +104,9 @@ class Hypre(AutotoolsPackage, CudaPackage):
     # Option added in v2.16.0
     conflicts('+mixedint', when='@:2.15')
 
+    # HIP support added in v2.21.0
+    conflicts('+rocm', when='@:2.21.0')
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/hypre/package.py]: Given HIP support added in v2.21.0, you'll want to change the `when` clause here:

```suggestion
    conflicts('+rocm', when='@:2.20')
```

**Review by tldahlgren — CHANGES_REQUESTED** [review]
I have not yet been able to build the package -- due to issues building dependencies -- on a Cray system at LLNL so cannot yet confirm the change.  However, there is a mismatch between the comment and one of the (new) conflicts that I figured is worth noting.

File: var/spack/repos/builtin/packages/hypre/package.py (hunk 1 above)
Comment (osborn9) [reply var/spack/repos/builtin/packages/hypre/package.py]: Thanks @tldahlgren for the catch!

**Comment by tldahlgren** [comment]
@spackbot run pipeline

File: var/spack/repos/builtin/packages/hypre/package.py [hunk 2]
```diff
@@ -184,6 +190,22 @@ def configure_args(self):
             ])
             if '@:2.20.99' in spec:
                 configure_args.append('--disable-cub')
+        if '+rocm' in spec:
+            configure_args.extend([
+                '--with-hip',
+                '--enable-rocrand',
+                '--enable-rocsparse'
```
Comment (balay) [inline var/spack/repos/builtin/packages/hypre/package.py:197]: I would think you might need:

```
    depends_on('hip', when='+rocm')
    depends_on('rocrand', when='+rocm')
    depends_on('rocsparse', when='+rocm')
```

And if  spack were to install these packages - their location won't be the same as `hip`  so somehow the appropriate includes/libs need to be passed in.

```console
[balay@instinct spack]$ spack find -p hip rocrand rocsparse
==> 3 installed packages
-- linux-centos7-broadwell / gcc@10.2.0 -------------------------
hip@4.3.1        /storage/users/balay/spack/opt/spack/linux-centos7-broadwell/gcc-10.2.0/hip-4.3.1-2skkbesaligdhmc2bx4p4cszjnukueyc
rocrand@4.3.1    /storage/users/balay/spack/opt/spack/linux-centos7-broadwell/gcc-10.2.0/rocrand-4.3.1-cp64x43alpb6vomimopg2z3foitncwul
rocsparse@4.3.1  /storage/users/balay/spack/opt/spack/linux-centos7-broadwell/gcc-10.2.0/rocsparse-4.3.1-uo2ptfp6akukt3kvfumqt352tmeq4aet
```
My changes (for these issues) for PETSc are at https://github.com/spack/spack/pull/27779

Comment (osborn9) [reply var/spack/repos/builtin/packages/hypre/package.py:197]: Thanks @balay! Currently, we're assuming `rocrand` and `rocsparse` are located where `hip` is installed, so we'd need to make the change to allow this within hypre.

**Comment by osborn9** [comment]
#29147 contains these changes, plus updates to CUDA variant and better handling of specifying gpu-arch.

**Pull request closed without merging**